exonum-java-binding/core/src/test/java/com/exonum/binding/core/storage/indices/ProofMapIndexProxyNoKeyHashingIntegrationTest.java

```diff
@@ -0,0 +1,160 @@
+  )
+      .map(ProofMapIndexProxyNoKeyHashingIntegrationTest::createRawProofKey)
+      .sorted(UnsignedBytes.lexicographicalComparator())
+      .map(HashCode::fromBytes)
+      .collect(Collectors.toList());
+
+  private static final HashCode INVALID_PROOF_KEY = HashCode.fromString("1234");
+
+  @Override
+  List<HashCode> getProofKeys() {
+    return PROOF_KEYS;
+  }
+
+  @Override
+  ProofMapIndexProxy<HashCode, String> create(String name, View view) {
+    return createProofMap(name, view);
+  }
+
+  @Override
+  ProofMapIndexProxy<HashCode, String> createInGroup(String groupName, byte[] idInGroup,
+                                                     View view) {
+    return ProofMapIndexProxy.newInGroupUnsafeNoKeyHashing(groupName, idInGroup, view,
+        StandardSerializers.hash(), StandardSerializers.string());
+  }
+
+  private static ProofMapIndexProxy<HashCode, String> createProofMap(String name, View view) {
+    return ProofMapIndexProxy.newInstanceNoKeyHashing(name, view, StandardSerializers.hash(),
+        StandardSerializers.string());
+  }
+
+  @Test
+  void containsThrowsIfInvalidKey() {
+    runTestWithView(database::createSnapshot, (map) -> assertThrows(IllegalArgumentException.class,
+        () -> map.containsKey(INVALID_PROOF_KEY)));
+  }
+
+  @Test
+  void putFailsIfInvalidKey() {
+    runTestWithView(database::createFork, (map) -> assertThrows(IllegalArgumentException.class,
+        () -> map.put(INVALID_PROOF_KEY, V1)));
+  }
+
+  @Test
+  void removeFailsIfInvalidKey() {
+    runTestWithView(database::createFork,
+        (map) -> assertThrows(IllegalArgumentException.class, () -> map.remove(INVALID_PROOF_KEY)));
+  }
+
+  @Test
+  void keysTest() {
+    runTestWithView(database::createFork, (map) -> {
+      List<MapEntry<HashCode, String>> entries = createSortedMapEntries();
+
+      putAll(map, entries);
+
+      Iterator<HashCode> keysIterator = map.keys();
+      List<HashCode> keysFromIter = ImmutableList.copyOf(keysIterator);
+      List<HashCode> keysInMap = MapEntries.extractKeys(entries);
+
+      // Keys must appear in a lexicographical order.
+      assertThat(keysFromIter, equalTo(keysInMap));
+    });
+  }
+
+  @Test
+  void valuesTest() {
+    runTestWithView(database::createFork, (map) -> {
+      List<MapEntry<HashCode, String>> entries = createSortedMapEntries();
+
+      putAll(map, entries);
+
+      Iterator<String> valuesIterator = map.values();
+      List<String> valuesFromIter = ImmutableList.copyOf(valuesIterator);
+      List<String> valuesInMap = MapEntries.extractValues(entries);
+
+      // Values must appear in a lexicographical order of keys.
+      assertThat(valuesFromIter, equalTo(valuesInMap));
+    });
+  }
+
+  @Test
+  void entriesTest() {
+    runTestWithView(database::createFork, (map) -> {
+      List<MapEntry<HashCode, String>> entries = createSortedMapEntries();
+
+      putAll(map, entries);
+
+      Iterator<MapEntry<HashCode, String>> entriesIterator = map.entries();
```

Comment:
(keys, values, entries) have incomplete documentation of the order of the elements missing the default variant.

Reply:
The conditional documentation (with the fact that the API users always need to know if hashing happens or not to access the map from other services or verify the proofs it produces) makes me wonder if we shall add two separate types. But that's certainly for a separate issue.

Reply:
❕